# Pandas. Домашнее задание

В домашнем задании есть обязательные и дополнительные задачи.

*Обязательные задачи* нужно сделать, чтобы проверить, что вы действительно усвоили материал модуля. Они обязательны к загрузке на проверку преподавателю.

*Дополнительные задачи* для тех, кто чувствует, что нужно ещё немного потренироваться и порешать задачки по работе с Pandas. Вы также можете загрузить их на проверку преподавателю и получить по ним обратную связь.

Удачи!



---



---



**Цели домашнего задания:**
1. Научиться работать с описательными статистиками набора данных.
2. Научиться выбирать данные по условиям.
3. Научиться объединять несколько датафреймов в один.


**Что входит в домашнее задание:**
1. *Обязательные задачи*:

17 задач с использованием:
- набора рекламных данных, показывающих, щелкнул ли конкретный пользователь интернета рекламу на веб-сайте компании или нет,
- дополнительного набора данных, описывающего пользователей, посещающих веб-сайт компании.
2. *Дополнительные задачи*:

3 задачи для отработки исследования данных.


**Критерии оценки:**
- программа выдаёт верный ответ на заданном наборе данных;
- код читабелен: переменным даны осмысленные названия, отступы и правила расстановки пробелов соблюдены.

**Как отправить задание на проверку**

Выполните задачи, сохраните изменения при помощи опции Save and Checkpoint из вкладки меню File или кнопки Save and Checkpoint на панели инструментов. Отправьте через форму на платформе итоговый файл Jupyter Notebook (в формате .ipynb) или ссылку на него.



---


---



**Обязательные задачи**

В течение всего домашнего задания вы будете работать с несколькими наборами данных:
1. Набор рекламных данных, показывающих, щёлкнул ли конкретный пользователь рекламу на сайте компании или нет. Состоит из трёх разных файлов: `advertising_1.csv`, `advertising_2.csv`, `advertising_3.csv`.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Daily Time Spent on Site**: Среднее ежедневное время нахождения пользователя на сайте в минутах.
- **Daily Internet Usage**: Среднее ежедневное время нахождения пользователя в интернете в минутах.
- **Ad Topic Line**: Заголовок рекламного объявления.
- **Clicked on Ad**: Кликнул пользователь на рекламу или нет.

Дополнительный к рекламным данным датасет, который описывает пользователей, посещающих сайт компании. Состоит из одного файла: users.csv.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Age**: Возраст пользователя.
- **Male**: Является ли пользователь мужчиной.
- **City**: Город пользователя.
- **Country**: Страна пользователя.
- **Area Income**: Средний доход в месте жительства пользователя.


***Задача 1***

Загрузите таблицу из файла `advertising_1.csv` и сохраните её в датафрейм `adv1_df`. Укажите в качестве индекса столбец `Number`. Выведите на экран первые десять строк.


In [3]:
import pandas as pd

In [4]:
# Ваш код здесь
adv1_df = pd.read_csv('advertising_1.csv', index_col='Number')
adv1_df = pd.DataFrame(adv1_df)
adv1_df.head(10)

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
11                         88.91                208.36   
12                           NaN                131.76   
13                         74.53                   NaN   
14                         69.88                183.82   
15                         47.64                122.02   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
11                 Enhanced dedicated support              0  
12                   Reactive local challenge              1  
13             Configurable coherent function              0  
14         Mandatory homogeneous architecture              0  
15             Centralized neutral neural-net              1

***Задача 2***

Выведите размер датафрейма `adv1_df`.

In [5]:
# Ваш код здесь
adv1_df.shape

(500, 4)

***Задача 3***

Выведите cреднее ежедневное время нахождения в интернете пользователя под номером 8.

In [6]:
# Ваш код здесь
adv1_df.loc[8, 'Daily Internet Usage']

245.89

***Задача 4***

Загрузите таблицу из файла `advertising_2.csv` и сохраните её в датафрейм `adv2_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран строки для пользователей с номерами с 533-го по 536-й.

In [7]:
# Ваш код здесь
adv2_df = pd.read_csv('advertising_2.csv', index_col='Number')
adv2_df = pd.DataFrame(adv2_df)
adv2_df.loc[533: 536]

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
533                        46.61                   NaN   
534                          NaN                162.03   
535                        65.10                118.10   
536                        53.44                   NaN   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
533             Ergonomic methodical encoding              1  
534         Compatible dedicated productivity              0  
535            Up-sized real-time methodology              1  
536     Up-sized next generation architecture              1

***Задача 5***

Выведите сводную таблицу описательных статистик для датафрейма `adv2_df`, вызвав необходимый метод на всём датафрейме. 

Глядя на вывод, напишите, столбец с каким названием не содержит незаполненных значений.

*Подсказка:*  В сводной таблице не содержится информации о количестве строк всего датафрейма, поэтому для сначала распечатайте общее количество строк датафрейма.

*Рекомендация:* Кроме изученного метода `describe` для описания датафрейма и получения информации о пропущенных значениях полезен метод [`info`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html).

In [15]:
# Ваш код здесь
adv2_df.describe()

Daily Time Spent on Site  Daily Internet Usage  Clicked on Ad
count                184.000000            184.000000     200.000000
mean                  66.494728            186.809674       0.420000
std                   15.414725             43.844580       0.494797
min                   32.910000            108.030000       0.000000
25%                   52.805000            148.955000       0.000000
50%                   71.150000            194.170000       0.000000
75%                   78.352500            225.372500       1.000000
max                   91.430000            259.760000       1.000000

In [16]:
adv2_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 200 entries, 530 to 739
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Daily Time Spent on Site  184 non-null    float64
 1   Daily Internet Usage      184 non-null    float64
 2   Ad Topic Line             200 non-null    object 
 3   Clicked on Ad             200 non-null    int64  
dtypes: float64(2), int64(1), object(1)
memory usage: 15.9+ KB


In [17]:
adv2_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
530                          NaN                165.27   
531                        43.65                   NaN   
532                        59.61                198.45   
533                        46.61                   NaN   
534                          NaN                162.03   
...                          ...                   ...   
735                        65.53                190.17   
736                        61.87                   NaN   
737                        83.16                   NaN   
738                        44.11                121.24   
739                        56.57                131.98   

                                       Ad Topic Line  Clicked on Ad  
Number                                                               
530                 Re-engineered impactful software              0  
531              Business-focused background synergy              1  
532     Future-proofed coherent budgetary management              0  
533                    Ergonomic methodical encoding              1  
534                Compatible dedicated productivity              0  
...                                              ...            ...  
735           Pre-emptive client-server installation              0  
736        Multi-channeled attitude-oriented toolset              0  
737                    Decentralized 24hour approach              0  
738                   Organic next generation matrix              1  
739             Multi-channeled non-volatile website              1  

[200 rows x 4 columns]

In [18]:
# Ваш код здесь
# Ваш код здесь
print("Не содержат незаполненных значений столбцы: 'Ad Topic Line' и 'Clicked on Ad'")
# Я не очень понял относиться ли Clicked on Ad к незаполненным, 
# но по моей логике 0 как 1 это данные о том что нажал пользователь на рекламу или нет
# так что я добавил его в список

Не содержат незаполненных значений столбцы: 'Ad Topic Line' и 'Clicked on Ad'


In [19]:
# или можно так

non_nan_column = adv2_df.columns[adv2_df.count() == len(adv2_df)].tolist()

print(f"Не содержат незаполненных значений столбцы: {non_nan_column}")

Не содержат незаполненных значений столбцы: ['Ad Topic Line', 'Clicked on Ad']


***Задача 6***

Добавьте к датафрейму `adv1_df` в конец датафрейм `adv2_df` с помощью метода `append`. 

Сохраните объединенный датафрейм в новой переменной `adv12_df` и распечатайте его.

In [20]:
# Ваш код здесь
adv12_df = adv1_df.append(adv2_df, ignore_index=True)
# буду использовать concat 
# ибо https://stackoverflow.com/questions/75956209/error-dataframe-object-has-no-attribute-append

AttributeError: 'DataFrame' object has no attribute 'append'

In [95]:
adv12_df = pd.concat([adv1_df, adv2_df])
adv12_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
735                        65.53                190.17   
736                        61.87                   NaN   
737                        83.16                   NaN   
738                        44.11                121.24   
739                        56.57                131.98   

                                    Ad Topic Line  Clicked on Ad  
Number                                                            
5              Cloned 5thgeneration orchestration              0  
6              Monitored national standardization              0  
8           Triple-buffered reciprocal time-frame              0  
9                   Robust logistical utilization              0  
10                Sharable client-driven software              0  
...                                           ...            ...  
735        Pre-emptive client-server installation              0  
736     Multi-channeled attitude-oriented toolset              0  
737                 Decentralized 24hour approach              0  
738                Organic next generation matrix              1  
739          Multi-channeled non-volatile website              1  

[700 rows x 4 columns]

***Задача 7***

Для объединённого датафрейма `adv12_df` распечатайте среднее время, проводимое пользователем на сайте компании.

In [22]:
# Ваш код здесь
adv12_df['Daily Time Spent on Site'].mean()

65.57268138801263

***Задача 8***

Из объединённого датафрейма `adv12_df` выберите и распечатайте строки только с теми пользователями, для которых неизвестно среднее ежедневное время, проводимое пользователем на сайте компании.

In [23]:
# Ваш код здесь
adv12_df[adv12_df['Daily Time Spent on Site'].isnull()]

Daily Time Spent on Site  Daily Internet Usage  \
3                         NaN                225.58   
6                         NaN                131.76   
13                        NaN                143.56   
20                        NaN                164.83   
27                        NaN                209.82   
..                        ...                   ...   
636                       NaN                235.97   
638                       NaN                126.44   
647                       NaN                200.71   
652                       NaN                226.69   
691                       NaN                225.05   

                               Ad Topic Line  Clicked on Ad  
3              Robust logistical utilization              0  
6                   Reactive local challenge              1  
13             Grass-roots coherent extranet              1  
20     Mandatory disintermediate utilization              1  
27   Business-focused value-added definition              0  
..                                       ...            ...  
636         Customizable value-added project              0  
638             Reactive impactful challenge              1  
647            Focused multi-state workforce              0  
652            Total coherent superstructure              0  
691          Multi-tiered heuristic strategy              0  

[66 rows x 4 columns]

***Задача 9***

Загрузите таблицу из файла `advertising_3.csv` и сохраните её в датафрейм `adv3_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран столбцы `Ad Topic Line` и `Clicked on Ad`.

In [85]:
# Ваш код здесь
adv3_df = pd.read_csv('advertising_3.csv', index_col='Number')
adv3_df = pd.DataFrame(adv3_df)
adv3_df[['Ad Topic Line', 'Clicked on Ad']]

Ad Topic Line  Clicked on Ad
Number                                                             
740                 Distributed bifurcated challenge              0
741       Customizable zero-defect Internet solution              0
742     Self-enabling zero administration neural-net              1
744           Open-architected system-worthy ability              0
745               Quality-focused maximized extranet              1
...                                              ...            ...
999                    Front-line bifurcated ability              1
1000                   Fundamental modular algorithm              1
1001                 Grass-roots cohesive monitoring              1
1002                    Expanded intangible solution              1
1003            Proactive bandwidth-monitored policy              0

[251 rows x 2 columns]

***Задача 10***

Добавьте к датафрейму `adv12_df` в конец датафрейм `adv3_df` с помощью метода `concat`. Явно пропишите значение параметра `axis`.

Сохраните объединённый датафрейм в новой переменной `adv123_df` и распечатайте его.

In [96]:
# Ваш код здесь
adv123_df = pd.concat([adv12_df, adv3_df], axis=0)
adv123_df

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
999                        43.70                173.01   
1000                       72.97                208.58   
1001                       51.30                   NaN   
1002                       51.63                120.37   
1003                       55.55                187.95   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
...                                       ...            ...  
999             Front-line bifurcated ability              1  
1000            Fundamental modular algorithm              1  
1001          Grass-roots cohesive monitoring              1  
1002             Expanded intangible solution              1  
1003     Proactive bandwidth-monitored policy              0  

[951 rows x 4 columns]

***Задача 11***

Для объединённого датафрейма `adv123_df` выберите только те строки, в которых пользователи кликнули на рекламное сообщение. Сохраните полученный результат в датафрейм `success_adv_df`, распечатайте размер нового датафрейма.

In [26]:
# Ваш код здесь
success_adv_df = adv123_df[adv123_df['Clicked on Ad'] == 1]
success_adv_df

Daily Time Spent on Site  Daily Internet Usage  \
6                         NaN                131.76   
9                       47.64                122.02   
11                      69.57                113.12   
13                        NaN                143.56   
14                      63.45                140.64   
..                        ...                   ...   
684                     67.59                113.69   
686                     43.07                137.63   
687                     39.47                163.48   
698                     44.11                121.24   
699                     56.57                131.98   

                                   Ad Topic Line  Clicked on Ad  
6                       Reactive local challenge              1  
9                 Centralized neutral neural-net              1  
11         Centralized content-based focus group              1  
13                 Grass-roots coherent extranet              1  
14            Persistent demand-driven interface              1  
..                                           ...            ...  
684     User-centric solution-oriented emulation              1  
686  Monitored zero administration collaboration              1  
687        Team-oriented systematic installation              1  
698               Organic next generation matrix              1  
699         Multi-channeled non-volatile website              1  

[336 rows x 4 columns]

***Задача 12***

Загрузите таблицу с данными пользователей из файла `users.csv` и сохраните её в датафрейм `users_df`. Индекс не указывайте.

Распечатайте сам датафрейм и сводную таблицу со статистиками для него.

In [87]:
# Ваш код здесь
users_df = pd.read_csv('users.csv')
users_df = pd.DataFrame(users_df)
users_df.describe()

Number         Age        Male   Area Income
count   980.000000  860.000000  868.000000    903.000000
mean    504.485714   36.073256    0.500000  55010.631938
std     287.956480    8.795992    0.500288  13423.710892
min       5.000000   19.000000    0.000000  13996.500000
25%     255.750000   29.000000    0.000000  46947.570000
50%     503.500000   35.000000    0.500000  57032.360000
75%     753.250000   42.000000    1.000000  65537.990000
max    1004.000000   61.000000    1.000000  79484.800000

***Задача 13***

Глядя на статистики `users_df`, укажите значения, удовлетворяющие следующим условиям:
1. Возраст самого молодого пользователя сайта: 19
2. Максимальный средний доход в месте жительства пользователя сайта: 79484.8
3. Средний возраст пользователя сайта: 36
4. Количество заполненных значений для колонки, содержащей информацию о том, является пользователь мужчиной или нет: 868
5. Минимальный номер пользователя: 5

В этом задании не требуется выбирать по условиям и распечатывать значения, достаточно внимательно посмотреть на статистики и скопировать нужные значения согласно вопросам.

***Задача 14***

К полученному датафрейму с успешными рекламными сообщениями `success_adv_df` добавьте справа информацию из датафрейма с пользователями `users_df` и сохраните в датафрейм `success_full_df`. В полученном датафрейме должны остаться только те пользователи, номера которых есть в обоих датафреймах. Для этого воспользуйтесь методом `merge` и пропишите явно верные параметры.

Затем распечатайте датафрейм.

*Важно:*

Объединять датафреймы с помощью `merge` можно и по индексам, и по столбцам с разными названиями. 
Так как общие значения содержатся в столбце `Number`, но в датафрейме `users_df` данный столбец является обычным столбцом, а в `success_adv_df` является индексом, перед тем как объединять, переведите столбец `Number` в `success_adv_df` в обычный столбец.

Для этого перед объединением пропишите:
```
success_adv_df = success_adv_df.reset_index()
```

[Подробнее о reset_index](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html)

In [28]:
# Ваш код здесь
success_adv_df = success_adv_df.reset_index(level=None)

In [97]:
success_adv_df = success_adv_df.rename(columns={'index': 'Number'})

success_adv_df

Number  Daily Time Spent on Site  Daily Internet Usage  \
0         6                       NaN                131.76   
1         9                     47.64                122.02   
2        11                     69.57                113.12   
3        13                       NaN                143.56   
4        14                     63.45                140.64   
..      ...                       ...                   ...   
331     684                     67.59                113.69   
332     686                     43.07                137.63   
333     687                     39.47                163.48   
334     698                     44.11                121.24   
335     699                     56.57                131.98   

                                   Ad Topic Line  Clicked on Ad  
0                       Reactive local challenge              1  
1                 Centralized neutral neural-net              1  
2          Centralized content-based focus group              1  
3                  Grass-roots coherent extranet              1  
4             Persistent demand-driven interface              1  
..                                           ...            ...  
331     User-centric solution-oriented emulation              1  
332  Monitored zero administration collaboration              1  
333        Team-oriented systematic installation              1  
334               Organic next generation matrix              1  
335         Multi-channeled non-volatile website              1  

[336 rows x 5 columns]

In [30]:
success_full_df = pd.merge(success_adv_df, users_df, how='inner', left_on='Number', right_index=True)

success_full_df

Number  Number_x  Daily Time Spent on Site  Daily Internet Usage  \
0         6         6                       NaN                131.76   
1         9         9                     47.64                122.02   
2        11        11                     69.57                113.12   
3        13        13                       NaN                143.56   
4        14        14                     63.45                140.64   
..      ...       ...                       ...                   ...   
331     684       684                     67.59                113.69   
332     686       686                     43.07                137.63   
333     687       687                     39.47                163.48   
334     698       698                     44.11                121.24   
335     699       699                     56.57                131.98   

                                   Ad Topic Line  Clicked on Ad  Number_y  \
0                       Reactive local challenge              1        11   
1                 Centralized neutral neural-net              1        15   
2          Centralized content-based focus group              1        18   
3                  Grass-roots coherent extranet              1        20   
4             Persistent demand-driven interface              1        21   
..                                           ...            ...       ...   
331     User-centric solution-oriented emulation              1       702   
332  Monitored zero administration collaboration              1       704   
333        Team-oriented systematic installation              1       705   
334               Organic next generation matrix              1       716   
335         Multi-channeled non-volatile website              1       717   

      Age  Male             City                 Country  Area Income  
0    33.0   0.0      Brandonstad                 Myanmar     53852.85  
1    49.0   NaN  West Brandonton                   Qatar     45632.51  
2    24.0   0.0              NaN  Bosnia and Herzegovina     51739.63  
3    23.0   NaN              NaN                   Spain     52182.23  
4     NaN   0.0   West Dylanberg   Palestinian Territory     23936.86  
..    ...   ...              ...                     ...          ...  
331  25.0   0.0   West Wendyland               Greenland     55195.61  
332  29.0   NaN      Kennethview            Saint Helena          NaN  
333   NaN   1.0   West Mariafort                 Lebanon     44490.09  
334  31.0   1.0        New Kayla             Netherlands     73413.87  
335   NaN   1.0      Carsonshire               Gibraltar     58114.30  

[336 rows x 12 columns]

In [31]:
success_outer_df = pd.merge(success_adv_df, users_df, how='outer', left_on='Number', right_index=True)

success_outer_df

Number  Number_x  Daily Time Spent on Site  Daily Internet Usage  \
0.0       6       6.0                       NaN                131.76   
1.0       9       9.0                     47.64                122.02   
2.0      11      11.0                     69.57                113.12   
3.0      13      13.0                       NaN                143.56   
4.0      14      14.0                     63.45                140.64   
..      ...       ...                       ...                   ...   
NaN     975       NaN                       NaN                   NaN   
NaN     976       NaN                       NaN                   NaN   
NaN     977       NaN                       NaN                   NaN   
NaN     978       NaN                       NaN                   NaN   
NaN     979       NaN                       NaN                   NaN   

                             Ad Topic Line  Clicked on Ad  Number_y   Age  \
0.0               Reactive local challenge            1.0        11  33.0   
1.0         Centralized neutral neural-net            1.0        15  49.0   
2.0  Centralized content-based focus group            1.0        18  24.0   
3.0          Grass-roots coherent extranet            1.0        20  23.0   
4.0     Persistent demand-driven interface            1.0        21   NaN   
..                                     ...            ...       ...   ...   
NaN                                    NaN            NaN      1000  30.0   
NaN                                    NaN            NaN      1001  45.0   
NaN                                    NaN            NaN      1002  51.0   
NaN                                    NaN            NaN      1003  19.0   
NaN                                    NaN            NaN      1004  26.0   

     Male             City                 Country  Area Income  
0.0   0.0      Brandonstad                 Myanmar     53852.85  
1.0   NaN  West Brandonton                   Qatar     45632.51  
2.0   0.0              NaN  Bosnia and Herzegovina     51739.63  
3.0   NaN              NaN                   Spain     52182.23  
4.0   0.0   West Dylanberg   Palestinian Territory     23936.86  
..    ...              ...                     ...          ...  
NaN   NaN        Duffystad                 Lebanon     71384.57  
NaN   1.0      New Darlene  Bosnia and Herzegovina     67782.17  
NaN   NaN    South Jessica                Mongolia     42415.72  
NaN   0.0      West Steven               Guatemala     41920.79  
NaN   0.0      Ronniemouth                  Brazil     29875.80  

[980 rows x 12 columns]

***Задача 15***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, отобразите сводную таблицу со статистиками для строковых колонок (`Ad Topic Line`, `City`, `Country`).


In [32]:
success_full_df[['Ad Topic Line', 'City', 'Country']].describe()

Ad Topic Line          City     Country
count                        336           300         308
unique                       336           298         187
top     Reactive local challenge  Williamsport  Micronesia
freq                           1             2           5

In [33]:
success_outer_df[['Ad Topic Line', 'City', 'Country']].describe()

Ad Topic Line       City Country
count                        336        887     888
unique                       336        860     235
top     Reactive local challenge  Lisamouth  France
freq                           1          3       9

***Задача 16***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите страну и количество вхождений этой страны в датафрейм.


In [34]:
# Ваш код здесь
success_full_df['Country'].value_counts()

Country
Micronesia               5
Taiwan                   5
Venezuela                4
Antigua and Barbuda      4
Fiji                     4
                        ..
Bulgaria                 1
Sao Tome and Principe    1
Saudi Arabia             1
United Arab Emirates     1
Gibraltar                1
Name: count, Length: 187, dtype: int64

In [35]:
success_outer_df['Country'].value_counts()

Country
France                   9
Czech Republic           9
Senegal                  8
Liberia                  8
Turkey                   8
                        ..
Saint Kitts and Nevis    1
Norway                   1
Jordan                   1
Djibouti                 1
Lesotho                  1
Name: count, Length: 235, dtype: int64

***Задача 17***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных.


In [36]:
# Ваш код здесь
success_full_df[success_full_df['Country'] == success_full_df['Country'].value_counts().idxmax()]

Number  Number_x  Daily Time Spent on Site  Daily Internet Usage  \
22       50        50                     50.33                133.20   
53      105       105                     63.24                235.78   
225     450       450                     47.48                141.34   
285     575       575                     50.18                127.82   
301     607       607                     61.04                149.21   

                              Ad Topic Line  Clicked on Ad  Number_y   Age  \
22              Sharable analyzing alliance              1        58  50.0   
53   Future-proofed holistic superstructure              1       114  32.0   
225      Reverse-engineered 24hour hardware              1       464   NaN   
285       Customizable hybrid system engine              1       590  46.0   
301   Operative didactic Local Area Network              1       623  39.0   

     Male                City     Country  Area Income  
22    1.0      South Lauraton  Micronesia     62657.53  
53    0.0            Johnstad  Micronesia     72272.90  
225   1.0            Jayville  Micronesia     51512.66  
285   1.0  Lake Elizabethside  Micronesia     41768.13  
301   0.0                 NaN  Micronesia     51593.46

In [37]:
success_outer_df[success_outer_df['Country'] == success_outer_df['Country'].value_counts().idxmax()]

Number  Number_x  Daily Time Spent on Site  Daily Internet Usage  \
134.0     248     248.0                     70.58                136.94   
159.0     308     308.0                     41.47                219.79   
233.0     466     466.0                     59.12                124.54   
NaN       611       NaN                       NaN                   NaN   
NaN       621       NaN                       NaN                   NaN   
NaN       756       NaN                       NaN                   NaN   
NaN       870       NaN                       NaN                   NaN   
NaN       957       NaN                       NaN                   NaN   
NaN       964       NaN                       NaN                   NaN   

                                   Ad Topic Line  Clicked on Ad  Number_y  \
134.0   Proactive next generation knowledge user            1.0       259   
159.0  Secured scalable Graphical User Interface            1.0       320   
233.0          Synergistic stable infrastructure            1.0       480   
NaN                                          NaN            NaN       628   
NaN                                          NaN            NaN       638   
NaN                                          NaN            NaN       775   
NaN                                          NaN            NaN       891   
NaN                                          NaN            NaN       982   
NaN                                          NaN            NaN       989   

        Age  Male            City Country  Area Income  
134.0  48.0   0.0    Port Crystal  France     62784.85  
159.0  39.0   0.0     North April  France     47160.53  
233.0  33.0   1.0       Smithtown  France     71296.67  
NaN    35.0   NaN  Port Brianfort  France     48918.55  
NaN    37.0   0.0       Adamsbury  France     55358.88  
NaN    30.0   0.0      Newmanberg  France     69758.31  
NaN     NaN   1.0      Mariahview  France     43073.78  
NaN     NaN   1.0    Robinsontown  France     51900.03  
NaN    31.0   1.0    Lindsaymouth  France     58037.66

***Задача 18***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных, они проводят в интернете в среднем больше 120 минут и их возраст меньше 30 лет. 
Cохраните получившиеся строки в новый датафрейм `filtered_df`.


In [38]:
# Ваш код здесь
most_common_country = success_full_df['Country'].value_counts().idxmax()

filtered_df1 = success_full_df[(success_full_df['Country'] == most_common_country) & 
                              (success_full_df['Daily Internet Usage'] > 120) & 
                              (success_full_df['Age'] < 30)]

filtered_df1


Empty DataFrame
Columns: [Number, Number_x, Daily Time Spent on Site, Daily Internet Usage, Ad Topic Line, Clicked on Ad, Number_y, Age, Male, City, Country, Area Income]
Index: []

In [39]:
most_common_country = success_outer_df['Country'].value_counts().idxmax()

filtered_outer_df = success_outer_df[(success_outer_df['Country'] == most_common_country) & 
                              (success_outer_df['Daily Internet Usage'] > 120) & 
                              (success_outer_df['Age'] < 30)]

filtered_outer_df


Empty DataFrame
Columns: [Number, Number_x, Daily Time Spent on Site, Daily Internet Usage, Ad Topic Line, Clicked on Ad, Number_y, Age, Male, City, Country, Area Income]
Index: []

In [40]:
success_full_df[success_full_df['Country'] == 'Micronesia'].info()

<class 'pandas.core.frame.DataFrame'>
Index: 5 entries, 22 to 301
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Number                    5 non-null      int64  
 1   Number_x                  5 non-null      int64  
 2   Daily Time Spent on Site  5 non-null      float64
 3   Daily Internet Usage      5 non-null      float64
 4   Ad Topic Line             5 non-null      object 
 5   Clicked on Ad             5 non-null      int64  
 6   Number_y                  5 non-null      int64  
 7   Age                       4 non-null      float64
 8   Male                      5 non-null      float64
 9   City                      4 non-null      object 
 10  Country                   5 non-null      object 
 11  Area Income               5 non-null      float64
dtypes: float64(5), int64(4), object(3)
memory usage: 520.0+ bytes


In [41]:
success_outer_df[success_outer_df['Country'] == 'Micronesia'].info()

<class 'pandas.core.frame.DataFrame'>
Index: 7 entries, 22.0 to nan
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Number                    7 non-null      int64  
 1   Number_x                  5 non-null      float64
 2   Daily Time Spent on Site  5 non-null      float64
 3   Daily Internet Usage      5 non-null      float64
 4   Ad Topic Line             5 non-null      object 
 5   Clicked on Ad             5 non-null      float64
 6   Number_y                  7 non-null      int64  
 7   Age                       6 non-null      float64
 8   Male                      7 non-null      float64
 9   City                      6 non-null      object 
 10  Country                   7 non-null      object 
 11  Area Income               7 non-null      float64
dtypes: float64(7), int64(2), object(3)
memory usage: 728.0+ bytes


In [42]:
success_full_df[success_full_df['Country'] == 'Taiwan']

Number  Number_x  Daily Time Spent on Site  Daily Internet Usage  \
120     224       224                     53.68                115.26   
151     285       285                     79.81                178.85   
158     305       305                     34.78                   NaN   
200     401       401                       NaN                120.75   
284     573       573                     71.83                135.48   

                           Ad Topic Line  Clicked on Ad  Number_y   Age  Male  \
120           Operative stable moderator              1       235  33.0   1.0   
151  Proactive context-sensitive project              1       296  32.0   0.0   
158     Exclusive even-keeled moratorium              1       317  22.0   NaN   
200          Total user-facing hierarchy              1       414  45.0   0.0   
284           Diverse background ability              1       588  33.0   0.0   

                   City Country  Area Income  
120          Turnerview  Taiwan     72707.87  
151       Greenechester  Taiwan     61610.05  
158          New Amanda  Taiwan     41851.38  
200  East Brittanyville  Taiwan     46693.76  
284          Tammyshire  Taiwan     68448.94

In [43]:
success_outer_df[success_outer_df['Country'] == 'Taiwan']

Number  Number_x  Daily Time Spent on Site  Daily Internet Usage  \
120.0     224     224.0                     53.68                115.26   
151.0     285     285.0                     79.81                178.85   
158.0     305     305.0                     34.78                   NaN   
200.0     401     401.0                       NaN                120.75   
284.0     573     573.0                     71.83                135.48   
NaN       880       NaN                       NaN                   NaN   
NaN       882       NaN                       NaN                   NaN   

                             Ad Topic Line  Clicked on Ad  Number_y   Age  \
120.0           Operative stable moderator            1.0       235  33.0   
151.0  Proactive context-sensitive project            1.0       296  32.0   
158.0     Exclusive even-keeled moratorium            1.0       317  22.0   
200.0          Total user-facing hierarchy            1.0       414  45.0   
284.0           Diverse background ability            1.0       588  33.0   
NaN                                    NaN            NaN       901   NaN   
NaN                                    NaN            NaN       903  37.0   

       Male                City Country  Area Income  
120.0   1.0          Turnerview  Taiwan     72707.87  
151.0   0.0       Greenechester  Taiwan     61610.05  
158.0   NaN          New Amanda  Taiwan     41851.38  
200.0   0.0  East Brittanyville  Taiwan     46693.76  
284.0   0.0          Tammyshire  Taiwan     68448.94  
NaN     1.0       Gonzalezburgh  Taiwan     62430.55  
NaN     0.0          Chrismouth  Taiwan     42078.89

In [44]:
success_tw_df[success_tw_df['Country'] == success_tw_df['Country'].value_counts().idxmax()]

NameError: name 'success_tw_df' is not defined

In [45]:
success_tw_df = success_full_df[success_full_df['Country'] != 'Micronesia']

most_common_country = success_tw_df['Country'].value_counts().idxmax()

filtered_df1 = success_tw_df[(success_tw_df['Country'] == most_common_country) & 
                              (success_tw_df['Daily Internet Usage'] > 120) & 
                              (success_tw_df['Age'] < 30)]

filtered_df1


Empty DataFrame
Columns: [Number, Number_x, Daily Time Spent on Site, Daily Internet Usage, Ad Topic Line, Clicked on Ad, Number_y, Age, Male, City, Country, Area Income]
Index: []

In [52]:
success_tw_outer_df[success_tw_outer_df['Country'] == success_tw_outer_df['Country'].value_counts().idxmax()]

Number  Number_x  Daily Time Spent on Site  Daily Internet Usage  \
134.0     248     248.0                     70.58                136.94   
159.0     308     308.0                     41.47                219.79   
233.0     466     466.0                     59.12                124.54   
NaN       611       NaN                       NaN                   NaN   
NaN       621       NaN                       NaN                   NaN   
NaN       756       NaN                       NaN                   NaN   
NaN       870       NaN                       NaN                   NaN   
NaN       957       NaN                       NaN                   NaN   
NaN       964       NaN                       NaN                   NaN   

                                   Ad Topic Line  Clicked on Ad  Number_y  \
134.0   Proactive next generation knowledge user            1.0       259   
159.0  Secured scalable Graphical User Interface            1.0       320   
233.0          Synergistic stable infrastructure            1.0       480   
NaN                                          NaN            NaN       628   
NaN                                          NaN            NaN       638   
NaN                                          NaN            NaN       775   
NaN                                          NaN            NaN       891   
NaN                                          NaN            NaN       982   
NaN                                          NaN            NaN       989   

        Age  Male            City Country  Area Income  
134.0  48.0   0.0    Port Crystal  France     62784.85  
159.0  39.0   0.0     North April  France     47160.53  
233.0  33.0   1.0       Smithtown  France     71296.67  
NaN    35.0   NaN  Port Brianfort  France     48918.55  
NaN    37.0   0.0       Adamsbury  France     55358.88  
NaN    30.0   0.0      Newmanberg  France     69758.31  
NaN     NaN   1.0      Mariahview  France     43073.78  
NaN     NaN   1.0    Robinsontown  France     51900.03  
NaN    31.0   1.0    Lindsaymouth  France     58037.66

In [53]:
success_tw_outer_df = success_outer_df[success_outer_df['Country'] != 'Micronesia']

most_common_country = success_tw_outer_df['Country'].value_counts().idxmax()

filtered_tw_outer_df = success_tw_outer_df[(success_tw_outer_df['Country'] == most_common_country) & 
                              (success_tw_outer_df['Daily Internet Usage'] > 120) & 
                              (success_tw_outer_df['Age'] < 30)]

filtered_tw_outer_df


Empty DataFrame
Columns: [Number, Number_x, Daily Time Spent on Site, Daily Internet Usage, Ad Topic Line, Clicked on Ad, Number_y, Age, Male, City, Country, Area Income]
Index: []

In [54]:
success_fr_outer_df = success_outer_df[success_outer_df['Country'] != 'France']

most_common_country = success_fr_outer_df['Country'].value_counts().idxmax()

filtered_fr_outer_df = success_fr_outer_df[(success_fr_outer_df['Country'] == most_common_country) & 
                              (success_fr_outer_df['Daily Internet Usage'] > 120) & 
                              (success_fr_outer_df['Age'] < 30)]

filtered_fr_outer_df


Empty DataFrame
Columns: [Number, Number_x, Daily Time Spent on Site, Daily Internet Usage, Ad Topic Line, Clicked on Ad, Number_y, Age, Male, City, Country, Area Income]
Index: []

In [55]:
success_outer_df[success_outer_df['Country'] == 'Spain']

Number  Number_x  Daily Time Spent on Site  Daily Internet Usage  \
3.0      13      13.0                       NaN                143.56   
NaN     457       NaN                       NaN                   NaN   
NaN     829       NaN                       NaN                   NaN   

                     Ad Topic Line  Clicked on Ad  Number_y   Age  Male  \
3.0  Grass-roots coherent extranet            1.0        20  23.0   NaN   
NaN                            NaN            NaN       471  35.0   0.0   
NaN                            NaN            NaN       850  42.0   1.0   

           City Country  Area Income  
3.0         NaN   Spain     52182.23  
NaN  Port Blake   Spain     25603.93  
NaN         NaN   Spain     61690.93

***Задача 19***

Для получившегося датафрейма `filtered_df` выведите строки с данными пользователей, для которых либо не указано среднее время, проведённое на сайте, либо указанное время больше 55 минут. Cохраните получившиеся строки в новый датафрейм `filtered_df2`.


In [58]:
# Ваш код здесь
filtered_df2 = filtered_df[(filtered_df['Daily Time Spent on Site'].isnull()) | 
                           (filtered_df['Daily Time Spent on Site'] > 55)]

NameError: name 'filtered_df' is not defined

**Дополнительные задачи**

***Задача 20***

Для датафрейма `adv123_df` выведите строки с данными, заголовок рекламного сообщения в которых включает слово `model`.

Для этого вам понадобится метод [`contains`](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html).

Cохраните получившиеся строки в новый датафрейм `model_df`.


In [73]:
# Ваш код здесь
model_df = adv123_df[adv123_df['Ad Topic Line'].str.contains('model', case=False)]

model_df

Daily Time Spent on Site  Daily Internet Usage  \
53                      44.98                129.31   
108                     69.00                   NaN   
185                     78.57                   NaN   
192                     76.90                212.67   
288                       NaN                161.16   
324                     65.77                   NaN   
471                     81.10                216.49   
508                     91.43                209.91   
594                     78.17                192.27   
672                     81.90                225.47   

                           Ad Topic Line  Clicked on Ad  
53                Organized global model              1  
108              Virtual composite model              0  
185        Exclusive client-driven model              0  
192        Synchronized multimedia model              0  
288           Extended interactive model              1  
324  Function-based fault-tolerant model              0  
471             Front-line dynamic model              0  
508             Organic motivating model              0  
594               Automated mobile model              0  
672  Multi-channeled 3rdgeneration model              0

***Задача 21***

Для получившегося датафрейма `model_df` выведите строки с данными пользователей, которые проводят на сайте больше среднего времени по всем пользователям в данном датафрейме. Cохраните получившиеся строки в новый датафрейм `model_popular_df`.


In [60]:
# Ваш код здесь
average_time = model_df['Daily Time Spent on Site'].mean()

model_popular_df = model_df[model_df['Daily Time Spent on Site'] > average_time]

model_popular_df

Daily Time Spent on Site  Daily Internet Usage  \
185                     78.57                   NaN   
192                     76.90                212.67   
471                     81.10                216.49   
508                     91.43                209.91   
594                     78.17                192.27   
672                     81.90                225.47   

                           Ad Topic Line  Clicked on Ad  
185        Exclusive client-driven model              0  
192        Synchronized multimedia model              0  
471             Front-line dynamic model              0  
508             Organic motivating model              0  
594               Automated mobile model              0  
672  Multi-channeled 3rdgeneration model              0

***Задача 22***

Для объединённого датафрейма `model_popular_df` выведите строки с данными пользователей, номера которых равны 501, 712 и 823. 

Для этого вам понадобится метод [`isin`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isin.html).

*Подсказка:* Не забудьте, что номера пользователей находятся не в обычной колонке, а в колонке, являющейся индексом, поэтому вам необходимо получить значения индекса, а не обычной колонки.

In [61]:
# Ваш код здесь
selected_users = model_popular_df[model_popular_df.index.isin([501, 712, 823])]

selected_users

Empty DataFrame
Columns: [Daily Time Spent on Site, Daily Internet Usage, Ad Topic Line, Clicked on Ad]
Index: []

In [67]:
model_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, 53 to 672
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Daily Time Spent on Site  9 non-null      float64
 1   Daily Internet Usage      7 non-null      float64
 2   Ad Topic Line             10 non-null     object 
 3   Clicked on Ad             10 non-null     int64  
dtypes: float64(2), int64(1), object(1)
memory usage: 400.0+ bytes


Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
712                         81.9                225.47   

                              Ad Topic Line  Clicked on Ad  
Number                                                      
712     Multi-channeled 3rdgeneration model              0